File: packages/contracts/src/types/order.ts
// CanonicalOrder — the normalized shape every channel (POS, QR, aggregator,
// direct) maps into. See docs/spec/ordering.md §Canonical order model and
// HOLLER_MASTER_PROMPT.md §16.
//
// Frozen at Milestone 0.5 (ADR-008). Money fields are integer paise
// (CLAUDE.md §Money) — never floating point.

import { z } from "zod";

export const OrderTypeSchema = z.enum([
  "DINE_IN",
  "TAKEAWAY",
  "DELIVERY",
  "AGGREGATOR",
  "QR",
  "ROOM_SERVICE",
  "CATERING",
]);
export type OrderType = z.infer<typeof OrderTypeSchema>;

// Channel of origin. Mirrors the CHECK on `order.source` in both stores; the
// two must agree member for member and `scripts/check-order-source-drift.mjs`
// fails the build when they do not.
//
// AGGREGATOR_ZOMATO and AGGREGATOR_SWIGGY are DEPRECATED (ADR-026): they name a
// platform in a schema that is otherwise platform-agnostic. Neither has ever
// been written. They stay because removing a CHECK member is breaking; the
// removal trigger is the next breaking contracts bump.
//
// AGGREGATOR and TABLE_TAB are the 0.8.1 additions and NOTHING WRITES EITHER
// YET — the aggregator accept path still writes DIRECT, and the table device
// (ADR-025) has no code at all.
export const OrderSourceSchema = z.enum([
  "POS",
  "QR",
  /** @deprecated ADR-026 — use "AGGREGATOR"; removal at the next breaking bump. */
  "AGGREGATOR_ZOMATO",
  /** @deprecated ADR-026 — use "AGGREGATOR"; removal at the next breaking bump. */
  "AGGREGATOR_SWIGGY",
  "DIRECT",
  "AGGREGATOR",
  "TABLE_TAB",
]);
export type OrderSource = z.infer<typeof OrderSourceSchema>;

// Order state machine — see docs/domain/ORDER_STATE_MACHINE.md. Do not add
// states here without updating that document and bumping schema_version.
export const OrderStatusSchema = z.enum([
  "DRAFT",
  "CONFIRMED",
  "SENT_TO_KITCHEN",
  "PREPARING",
  "READY",
  "SERVED",
  "BILLED",
  "PAID",
  "CLOSED",
  "CANCELLED",
]);
export type OrderStatus = z.infer<typeof OrderStatusSchema>;

// THE STATUSES IN WHICH AN ORDER'S LINES MAY STILL BE AMENDED. Added at
// 0.8.3 (ADR-028).
//
// Adding a line to an order that has already gone to the kitchen is the
// normal way a restaurant works — a table orders another round, a waiter
// sends a starter and a main separately — and the edge has allowed it since
// `#132-A` (docs/spec/kitchen.md). The cloud allowed it only in DRAFT and
// refused everything else with a permanent 409, which under an at-least-once
// outbox means the order's whole queue wedges behind the refusal.
//
// **THIS SET IS DECLARED ONCE AND CONSUMED, NEVER RESTATED.** It had three
// independent declarations — a Rust match arm at the edge, a Go equality test
// in the cloud, and one word of English in openapi.yaml — and they disagreed
// for months without anything going red, because the defect that hid it
// (orders stuck in DRAFT cloud-side, fixed the same day) meant the cloud rule
// was never reached. `scripts/check-order-amendable-drift.mjs` now fails the
// build if any surface disagrees with this array.
//
// The edge is the authority for order transactions (§50.1), so this is the
// EDGE's rule written down, not a negotiated middle: the cloud replays what
// the outlet did and does not get an amendable set of its own.
//
// READY, SERVED, BILLED, PAID, CLOSED and CANCELLED are deliberately absent.
// By then the correction belongs to a new order or an explicit reopen, not a
// silent edit of a bill already on its way to the guest.
export const ORDER_ITEM_AMENDABLE_STATUSES = [
  "DRAFT",
  "CONFIRMED",
  "SENT_TO_KITCHEN",
  "PREPARING",
] as const satisfies readonly OrderStatus[];

/** True when `status` still permits a line to be added, removed or resized. */
export function isOrderItemAmendable(status: OrderStatus): boolean {
  return (ORDER_ITEM_AMENDABLE_STATUSES as readonly OrderStatus[]).includes(status);
}

export const OrderItemModifierSchema = z.object({
  modifier_id: z.string().uuid(),
  group_name: z.string(),
  option_name: z.string(),
  price_delta_paise: z.number().int(),
});
export type OrderItemModifier = z.infer<typeof OrderItemModifierSchema>;

export const OrderItemSchema = z.object({
  id: z.string().uuid(),
  menu_item_id: z.string().uuid(),
  variant_id: z.string().uuid().nullable(),
  quantity: z.number().int().positive(),
  unit_price_paise: z.number().int(), // snapshot at order time — never recomputed from live menu
  line_total_paise: z.number().int(),
  modifiers: z.array(OrderItemModifierSchema).default([]),
  notes: z.string().nullable(),
});
export type OrderItem = z.infer<typeof OrderItemSchema>;

export const CanonicalOrderSchema = z.object({
  holler_order_id: z.string().uuid(),
  external_order_id: z.string().nullable(), // aggregator's own order id; null for POS/QR/Direct origin
  source: OrderSourceSchema,
  outlet_id: z.string().uuid(),

  // Short human-facing number ('#A184'), minted edge-side alongside the order.
  // Added at 0.4.0 (ADR-016) to close the M2 finding that a printed KOT
  // carried the raw UUID — a cook cannot read one aloud across a kitchen.
  // CLAUDE.md §Money/time/identifiers requires human-facing numbers be short
  // and forbids exposing sequential PKs as security identifiers, so this is a
  // display string, never a key.
  //
  // Nullable ONLY for rows written before 0.4.0: SQLite cannot add a NOT NULL
  // column to a populated table without rebuilding "order", and a rebuild of
  // a table that order_item, kot and invoice all reference is the worse risk.
  // Every create path populates it; readers fall back to the id for legacy
  // rows alone.
  display_number: z.string().nullable(),

  order_type: OrderTypeSchema,
  status: OrderStatusSchema,
  table_id: z.string().uuid().nullable(),

  customer: z
    .object({
      name: z.string().nullable(),
      phone: z.string().nullable(),
    })
    .nullable(),
  delivery_address: z.string().nullable(),

  items: z.array(OrderItemSchema),

  subtotal_paise: z.number().int(),
  discount_paise: z.number().int().default(0),
  packaging_paise: z.number().int().default(0),
  delivery_charge_paise: z.number().int().default(0),
  taxes_paise: z.number().int().default(0),
  aggregator_discount_paise: z.number().int().default(0),
  merchant_discount_paise: z.number().int().default(0),
  total_paise: z.number().int(),

  payment_status: z.enum(["UNPAID", "PARTIALLY_PAID", "PAID", "REFUNDED"]),
  payment_source: z.string().nullable(),

  preparation_time_minutes: z.number().int().nullable(),
  rider: z
    .object({ name: z.string(), phone: z.string(), status: z.string() })
    .nullable(),

  timestamps: z.object({
    created_at: z.string().datetime(),
    confirmed_at: z.string().datetime().nullable(),
    updated_at: z.string().datetime(),
  }),
  source_payload: z.record(z.unknown()).nullable(), // raw external payload, audit only — never parsed as core data

  schema_version: z.literal(1),
});
export type CanonicalOrder = z.infer<typeof CanonicalOrderSchema>;

// OrderCommand set — the only sanctioned way to mutate an order; enforces
// the state machine at the command layer (docs/domain/ORDER_STATE_MACHINE.md).
export const OrderCommandSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("CONFIRM_ORDER"), order_id: z.string().uuid() }),
  z.object({ type: z.literal("SEND_TO_KITCHEN"), order_id: z.string().uuid() }),
  z.object({ type: z.literal("MARK_READY"), order_id: z.string().uuid() }),
  z.object({ type: z.literal("MARK_SERVED"), order_id: z.string().uuid() }),
  z.object({ type: z.literal("BILL_ORDER"), order_id: z.string().uuid() }),
  z.object({
    type: z.literal("CANCEL_ORDER"),
    order_id: z.string().uuid(),
    reason: z.string(),
  }),

  // Milestone 3 additions (ADR-016).
  //
  // SET_ORDER_ITEM_QUANTITY is a single command by design. Quantity must NOT
  // be implemented as remove-then-add: that is two durable writes with a crash
  // window between them, which is precisely the loss the durable-cart work
  // eliminated (docs/backlog-m2.md, docs/retro.md 2026-08-10). One command,
  // one write.
  z.object({
    type: z.literal("SET_ORDER_ITEM_QUANTITY"),
    order_id: z.string().uuid(),
    order_item_id: z.string().uuid(),
    quantity: z.number().int().positive(),
  }),
  z.object({
    type: z.literal("APPLY_DISCOUNT"),
    order_id: z.string().uuid(),
    discount_definition_id: z.string().uuid(),
    // LINE scope targets one line; BILL scope leaves this null.
    order_item_id: z.string().uuid().nullable(),
    // Required when the discount definition sets requires_reason.
    reason: z.string().nullable(),
  }),
  z.object({
    type: z.literal("PAY_ORDER"),
    order_id: z.string().uuid(),
  }),
  z.object({
    type: z.literal("CLOSE_ORDER"),
    order_id: z.string().uuid(),
  }),
]);
export type OrderCommand = z.infer<typeof OrderCommandSchema>;

File: packages/contracts/go/order.go
// Package contracts mirrors packages/contracts/src/types/*.ts. Frozen at
// Milestone 0.5 (ADR-008); round-trip-tested against packages/contracts/fixtures
// by CI's contract-drift check. Money fields are integer paise (CLAUDE.md).
package contracts

import "time"

type OrderType string

const (
	OrderTypeDineIn      OrderType = "DINE_IN"
	OrderTypeTakeaway    OrderType = "TAKEAWAY"
	OrderTypeDelivery    OrderType = "DELIVERY"
	OrderTypeAggregator  OrderType = "AGGREGATOR"
	OrderTypeQR          OrderType = "QR"
	OrderTypeRoomService OrderType = "ROOM_SERVICE"
	OrderTypeCatering    OrderType = "CATERING"
)

type OrderSource string

const (
	OrderSourcePOS              OrderSource = "POS"
	OrderSourceQR               OrderSource = "QR"
	OrderSourceAggregatorZomato OrderSource = "AGGREGATOR_ZOMATO"
	OrderSourceAggregatorSwiggy OrderSource = "AGGREGATOR_SWIGGY"
	OrderSourceDirect           OrderSource = "DIRECT"
)

// OrderStatus mirrors docs/domain/ORDER_STATE_MACHINE.md. Do not add states
// here without updating that document and bumping SchemaVersion.
type OrderStatus string

const (
	OrderStatusDraft          OrderStatus = "DRAFT"
	OrderStatusConfirmed      OrderStatus = "CONFIRMED"
	OrderStatusSentToKitchen  OrderStatus = "SENT_TO_KITCHEN"
	OrderStatusPreparing      OrderStatus = "PREPARING"
	OrderStatusReady          OrderStatus = "READY"
	OrderStatusServed         OrderStatus = "SERVED"
	OrderStatusBilled         OrderStatus = "BILLED"
	OrderStatusPaid           OrderStatus = "PAID"
	OrderStatusClosed         OrderStatus = "CLOSED"
	OrderStatusCancelled      OrderStatus = "CANCELLED"
)

type PaymentStatus string

const (
	PaymentStatusUnpaid         PaymentStatus = "UNPAID"
	PaymentStatusPartiallyPaid  PaymentStatus = "PARTIALLY_PAID"
	PaymentStatusPaid           PaymentStatus = "PAID"
	PaymentStatusRefunded       PaymentStatus = "REFUNDED"
)

type OrderItemModifier struct {
	ModifierID      string `json:"modifier_id"`
	GroupName       string `json:"group_name"`
	OptionName      string `json:"option_name"`
	PriceDeltaPaise int64  `json:"price_delta_paise"`
}

type OrderItem struct {
	ID              string              `json:"id"`
	MenuItemID      string              `json:"menu_item_id"`
	VariantID       *string             `json:"variant_id"`
	Quantity        int                 `json:"quantity"`
	UnitPricePaise  int64               `json:"unit_price_paise"`
	LineTotalPaise  int64               `json:"line_total_paise"`
	Modifiers       []OrderItemModifier `json:"modifiers"`
	Notes           *string             `json:"notes"`
}

type OrderCustomer struct {
	Name  *string `json:"name"`
	Phone *string `json:"phone"`
}

type OrderRider struct {
	Name   string `json:"name"`
	Phone  string `json:"phone"`
	Status string `json:"status"`
}

type OrderTimestamps struct {
	CreatedAt   time.Time  `json:"created_at"`
	ConfirmedAt *time.Time `json:"confirmed_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// CanonicalOrder mirrors src/types/order.ts CanonicalOrderSchema. Every
// order-producing channel normalizes into this shape (docs/spec/ordering.md).
type CanonicalOrder struct {
	HollerOrderID    string      `json:"holler_order_id"`
	ExternalOrderID  *string     `json:"external_order_id"`
	Source           OrderSource `json:"source"`
	OutletID         string      `json:"outlet_id"`

	OrderType OrderType   `json:"order_type"`
	Status    OrderStatus `json:"status"`
	TableID   *string     `json:"table_id"`

	Customer          *OrderCustomer `json:"customer"`
	DeliveryAddress   *string        `json:"delivery_address"`

	Items []OrderItem `json:"items"`

	SubtotalPaise           int64 `json:"subtotal_paise"`
	DiscountPaise           int64 `json:"discount_paise"`
	PackagingPaise          int64 `json:"packaging_paise"`
	DeliveryChargePaise     int64 `json:"delivery_charge_paise"`
	TaxesPaise              int64 `json:"taxes_paise"`
	AggregatorDiscountPaise int64 `json:"aggregator_discount_paise"`
	MerchantDiscountPaise   int64 `json:"merchant_discount_paise"`
	TotalPaise              int64 `json:"total_paise"`

	PaymentStatus PaymentStatus `json:"payment_status"`
	PaymentSource *string       `json:"payment_source"`

	PreparationTimeMinutes *int        `json:"preparation_time_minutes"`
	Rider                  *OrderRider `json:"rider"`

	Timestamps    OrderTimestamps        `json:"timestamps"`
	SourcePayload map[string]interface{} `json:"source_payload"` // raw external payload, audit only

	SchemaVersion int `json:"schema_version"`
}

// OrderCommand mirrors the discriminated union in src/types/order.ts. Go
// lacks native discriminated unions, so Type selects which optional fields
// apply; callers switch on Type.
type OrderCommand struct {
	Type    string  `json:"type"` // CONFIRM_ORDER | SEND_TO_KITCHEN | MARK_READY | MARK_SERVED | BILL_ORDER | CANCEL_ORDER
	OrderID string  `json:"order_id"`
	Reason  *string `json:"reason,omitempty"` // CANCEL_ORDER only
}
